### PR TITLE
Add custom module ceph_pool

### PR DESCRIPTION
**Pull request by msambol**
This module automates the creation of the ceph pool and PG count.
PG count depends largely on the number of OSDs and the number of
copies.

### CODE REVIEW

File: library/ceph_pool.py [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+      - The pool in question: create it or ensure correct pg count
+    required: true
+  constant:
+    description:
+      - A constant multiplier used in calculating pg count
+    required: true
+  copies:
+    description:
+      - The number of copies the cluster keeps
+    required: true
```
Comment (j2sol) [inline library/ceph_pool.py]: So pg count can be corrected, but not number of copies? What if you supply a different number than what already exists?

Comment (msambol) [inline library/ceph_pool.py]: I think we will always 3 copies, regardless of the size of the cluster. However, if new osds are added, we need to increase the pg count.

File: library/ceph_pool.py [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+    )
+
+    pool_name = module.params.get('pool_name')
+    constant = int(module.params.get('constant'))
+    copies = int(module.params.get('copies'))
+
+    # determine number of osds in cluster
+    cmd = ['ceph', 'osd', 'stat']
+    rc, out, err = module.run_command(cmd, check_rc=True)
+    osds = int(out.splitlines()[0].split(":")[1].strip().split(" ")[0])
```
Comment (j2sol) [inline library/ceph_pool.py]: That's a lot of text manipulation. It'd help if you put a comment in with the expected input string and the expected output string, which will help should the command output ever change.

File: library/ceph_pool.py [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+
+    # calculate desired pg count
+    total_pg_count = (osds * constant) / copies
+    i = desired_pg_count = 0
```
Comment (j2sol) [inline library/ceph_pool.py]: This is uncommon syntax, might want to comment on what is being attempted here.

Comment (msambol) [inline library/ceph_pool.py]: Just setting everything to 0. I can move to separate lines.

File: library/ceph_pool.py [hunk 4]
```diff
@@ -0,0 +1,121 @@
...
+    if rc != 0:
+        # create the pool
+        cmd = ['ceph', 'osd', 'pool', 'create', pool_name,
+               str(desired_pg_count), str(desired_pg_count)]
+        rc, out, err = module.run_command(cmd, check_rc=True)
+        module.exit_json(changed=True, msg="new pool was created")
+    # yes
+    else:
+        # does the current pg count match the desired pg count?
+        current_pg_count = int(out.splitlines()[0].split(":")[1].strip())
```
Comment (j2sol) [inline library/ceph_pool.py]: Again may want to comment on what the expected input string is.

File: library/ceph_pool.py [hunk 5]
```diff
@@ -0,0 +1,121 @@
...
+def increase_pg_count(module, osds, pool_name,
+                      current_pg_count, desired_pg_count):
+    max_increase_per_pass = osds * 32
+    diff = desired_pg_count - current_pg_count
+    while diff > 0:
+        if diff <= max_increase_per_pass:
+            cmd = ['ceph', 'osd', 'pool', 'set', pool_name,
+                   'pg_num', str(desired_pg_count)]
+            rc, out, err = module.run_command(cmd, check_rc=True)
+            time.sleep(30)
```
Comment (j2sol) [inline library/ceph_pool.py]: 30 seconds is a long time, is that really expected?

Comment (msambol) [inline library/ceph_pool.py]: I hate it too. That's what they had in ICOS. Let me try to see exactly how long.

File: library/ceph_pool.py (hunk 1 above)
Comment (j2sol) [inline library/ceph_pool.py]: I don't understand what pg count and copies mean, so why would I be able to force a change of one but not the other?

Comment (msambol) [inline library/ceph_pool.py]: pg count stands for placement group. here's some more info how we calculate it: http://ceph.com/pgcalc/. copies is just the number of copies of the data we are keeping. basically a constant 3.

Comment (j2sol) [inline library/ceph_pool.py]: If it's a constant 3, why do I have to supply the option when calling the module? Could the argument default to 3?

Also, are you intending this to be pushed upstream to Ansible(-modules-extras) ?

Comment (msambol) [inline library/ceph_pool.py]: My intent was to grab the 3 from `ceph_common. pool_default_size`. 

And no I didn't plan to push it upstream..

Comment (j2sol) [inline library/ceph_pool.py]: Alright, from an outsider reading the module it looks odd that I can set one attribute and expect it to change, but I can't set another attribute and expect it to change. I don't know what to expect out of the module if I supply a number of copies that's different than what the pool already has.

Comment (msambol) [inline library/ceph_pool.py]: You have a point.. Here's how I called this in my testing. I could hardcode the 3 in three in there if you want?

```
- name: create openstack pool
  ceph_pool:
    pool_name: "{{ ceph_common.pool_name}}"
    constant: "{{ ceph_common.pool_constant }}"
    copies: "{{ ceph_common.pool_default_size }}"
  register: pool_output
  run_once: true
  delegate_to: "{{ groups[ceph.monitor_group_name][0] }}"
```

Comment (j2sol) [inline library/ceph_pool.py]: Well, I'm trying to look at the module without the context of how you'll be using it. I'm reviewing the module as a stand alone piece of work.

Perhaps the module should error if the supplied copies argument doesn't equal the number of copies a pool already has?

Comment (msambol) [inline library/ceph_pool.py]: yeah, it's a pretty "custom" module which I don't really like but I thought was the best way to do what we needed. 

I could add that check it:
```
root@qint-lon02-c4:~# ceph osd pool get default size
size: 3
```

**Comment by j2sol** [comment]
This looks fine now. Tests won't even touch this.

**Pull request merged**